crypto/cmp/cmp_ctx.c

```diff
@@ -0,0 +1,1276 @@
+    return NULL;
+}
+
+/*
+ * Copies any given stack of inbound X509 certificates to extraCertsIn of
+ * the OSSL_CMP_CTX structure so that they may be retrieved later.
+ * returns 1 on success, 0 on error
+ */
+int CMP_CTX_set1_extraCertsIn(OSSL_CMP_CTX *ctx,
+                                   STACK_OF(X509) *extraCertsIn)
+{
+    if (ctx == NULL)
+        goto err;
+    if (extraCertsIn == NULL)
+        return 1;
+
+    /* if there are already inbound extraCerts on the stack delete them */
+    if (ctx->extraCertsIn != NULL) {
+        sk_X509_pop_free(ctx->extraCertsIn, X509_free);
+        ctx->extraCertsIn = NULL;
+    }
+
+    if ((ctx->extraCertsIn = X509_chain_up_ref(extraCertsIn)) == NULL) {
+        CMPerr(CMP_F_CMP_CTX_SET1_EXTRACERTSIN, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+
+    return 1;
+ err:
+    CMPerr(CMP_F_CMP_CTX_SET1_EXTRACERTSIN, CMP_R_NULL_ARGUMENT);
+    return 0;
+}
+
+/*
+ * Duplicate and push the given X509 certificate to the stack of
+ * outbound certificates to send in the extraCerts field.
+ * returns 1 on success, 0 on error
+ */
+int OSSL_CMP_CTX_extraCertsOut_push1(OSSL_CMP_CTX *ctx, const X509 *val)
+{
+    if (ctx == NULL)
+        goto err;
+    if ((ctx->extraCertsOut == NULL
+             && (ctx->extraCertsOut = sk_X509_new_null()) == NULL)
+            || !sk_X509_push(ctx->extraCertsOut, X509_dup(val))) {
+        CMPerr(CMP_F_OSSL_CMP_CTX_EXTRACERTSOUT_PUSH1, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+    return 1;
+ err:
+    CMPerr(CMP_F_OSSL_CMP_CTX_EXTRACERTSOUT_PUSH1, CMP_R_INVALID_ARGS);
+    return 0;
+}
+
+/*
+ * Duplicate and set the given stack as the new stack of X509
+ * certificates to send out in the extraCerts field.
+ * returns 1 on success, 0 on error
+ */
+int OSSL_CMP_CTX_set1_extraCertsOut(OSSL_CMP_CTX *ctx,
+                                    STACK_OF(X509) *extraCertsOut)
+{
+    if (ctx == NULL)
+        goto err;
+    if (extraCertsOut == NULL)
+        goto err;
+
+    if (ctx->extraCertsOut) {
+        sk_X509_pop_free(ctx->extraCertsOut, X509_free);
+        ctx->extraCertsOut = NULL;
+    }
+
+    if ((ctx->extraCertsOut = X509_chain_up_ref(extraCertsOut)) == NULL) {
+        CMPerr(CMP_F_OSSL_CMP_CTX_SET1_EXTRACERTSOUT, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+
+    return 1;
+ err:
+    CMPerr(CMP_F_OSSL_CMP_CTX_SET1_EXTRACERTSOUT, CMP_R_NULL_ARGUMENT);
+    return 0;
+}
+
+/*
+ * OSSL_CMP_CTX_policyOID_push1() adds the certificate policy OID given by the
+ * string to the X509_EXTENSIONS of the requested certificate template.
+ * returns 1 on success, -1 on parse error, and 0 on other error.
+ */
+int OSSL_CMP_CTX_policyOID_push1(OSSL_CMP_CTX *ctx, const char *policyOID)
+{
+    ASN1_OBJECT *policy;
+    POLICYINFO *pinfo = NULL;
+
+    if ((ctx == NULL) || (policyOID == NULL))
+        goto err;
+
+    if ((policy = OBJ_txt2obj(policyOID, 1)) == 0)
+        return -1; /* parse error */
```

Comment:
Yes, we need it because it is good give users a hint whether they provided invalid input or there was an internal error. This distinction is used in the upcoming `apps/cmp.c`.